aptos-move/framework/aptos-stdlib/sources/cryptography/algebra.move

```diff
@@ -16,14 +48,154 @@ module aptos_std::algebra {
+        Element<S> {
+            handle: sqr_internal<S>(x.handle)
+        }
+    }
+
+    /// Try computing `x^(-1)` for an element `x` of a structure `S`.
+    /// Return none if `x` equals to `zero<S>()`.
+    public fun inv<S>(x: &Element<S>): Option<Element<S>> {
+        abort_unless_cryptography_algebra_natives_enabled();
+        let (succeeded, handle) = inv_internal<S>(x.handle);
+        if (succeeded) {
+            let scalar = Element<S> { handle };
+            some(scalar)
+        } else {
+            none()
+        }
+    }
+
+    /// Compute `2*P` for an element `P` of a structure `S`. Faster and cheaper than `P + P`.
+    public fun double<S>(element_p: &Element<S>): Element<S> {
+        abort_unless_cryptography_algebra_natives_enabled();
+        Element<S> {
+            handle: double_internal<S>(element_p.handle)
+        }
+    }
+
+    /// Compute `k[0]*P[0]+...+k[n-1]*P[n-1]`, where
+    /// `P[]` are `n` elements of group `G` represented by parameter `elements`, and
+    /// `k[]` are `n` elements of the scalarfield `S` of group `G` represented by parameter `scalars`.
+    ///
+    /// Abort with code 0x010000 if the sizes of `elements` and `scalars` do not match.
+    public fun multi_scalar_mul<G, S>(elements: &vector<Element<G>>, scalars: &vector<Element<S>>): Element<G> {
+        let element_handles = handles_from_elements(elements);
+        let scalar_handles = handles_from_elements(scalars);
+        Element<G> {
+            handle: multi_scalar_mul_internal<G, S>(element_handles, scalar_handles)
+        }
+    }
+
+    /// Compute `k*P`, where `P` is an element of a group `G` and `k` is an element of the scalar field `S` of a structure `G`.
+    public fun scalar_mul<G, S>(element_p: &Element<G>, scalar_k: &Element<S>): Element<G> {
+        abort_unless_cryptography_algebra_natives_enabled();
+        Element<G> {
+            handle: scalar_mul_internal<G, S>(element_p.handle, scalar_k.handle)
+        }
+    }
+
+    /// Efficiently compute `e(P[0],Q[0])+...+e(P[n-1],Q[n-1])`,
+    /// where `e: (G1,G2) -> (Gt)` is a pre-compiled pairing function from groups `(G1,G2)` to group `Gt`,
+    /// `P[]` are `n` elements of group `G1` represented by parameter `g1_elements`, and
+    /// `Q[]` are `n` elements of group `G2` represented by parameter `g2_elements`.
+    ///
+    /// Abort with code 0x010000 if the sizes of `g1_elements` and `g2_elements` do not match.
+    public fun multi_pairing<G1,G2,Gt>(g1_elements: &vector<Element<G1>>, g2_elements: &vector<Element<G2>>): Element<Gt> {
+        abort_unless_cryptography_algebra_natives_enabled();
+        let g1_handles = handles_from_elements(g1_elements);
+        let g2_handles = handles_from_elements(g2_elements);
+        Element<Gt> {
+            handle: multi_pairing_internal<G1,G2,Gt>(g1_handles, g2_handles)
+        }
+    }
+
+    /// Compute a pre-compiled pairing function (a.k.a., bilinear map) on `element_1` and `element_2`.
```

Comment:
Not sure whether "pre-compiled" servers any purpose here.

Reply:
Fixed.